### ISSUE TITLE
AstGen producing ZIR that is missing type coercion for runtime conditional branching

### ISSUE THREAD

**Issue by andrewrk** [ziglang/zig#10816]
### Zig Version

0.10.0-dev.566+4b3b48762

### Steps to Reproduce

```
export fn entry() void {
    var cond: bool = true;
    const x: u16 = if (cond) 0x8000 else 0;
    _ = @as(u16, x);
}
```

```
$ zig build-obj test.zig
```

### Expected Behavior

This file should compile with no problems.

### Actual Behavior

```
thread 438054 panic: reached unreachable code
Analyzing test2.zig: test2.zig:entry
      %3 = dbg_stmt(2, 5)
      %4 = alloc_mut(@Ref.bool_type) 
      %5 = store_node(%4, @Ref.bool_true) 
      %6 = dbg_stmt(3, 5)
      %11 = block({
        %8 = load(%4) 
        %9 = as_node(@Ref.bool_type, %8) 
        %10 = condbr(%9, {
          %12 = int(32768)
          %15 = break(%11, %12)
        }, {
          %16 = break(%11, @Ref.zero)
        }) 
      }) 
      %18 = dbg_stmt(4, 5)
    > %19 = as_node(@Ref.u16_type, %11) 
      %20 = ensure_result_non_error(%19) 
      %21 = ret_coerce(@Ref.void_value) 
    For full context, use the command
      zig ast-check -t test2.zig


/home/andy/misc/zig/src/type.zig:3170:21: 0x2ee10ea in type.Type.intInfo (zig)
            else => unreachable,
                    ^
/home/andy/misc/zig/src/Sema.zig:14184:49: 0x3481d7d in Sema.coerce (zig)
                const src_info = inst_ty.intInfo(target);
                                                ^
/home/andy/misc/zig/src/Sema.zig:5794:23: 0x34836e4 in Sema.analyzeAs (zig)
    return sema.coerce(block, dest_ty, operand, src);
                      ^
/home/andy/misc/zig/src/Sema.zig:5782:26: 0x332b613 in Sema.zirAsNode (zig)
    return sema.analyzeAs(block, src, extra.dest_type, extra.operand);
                         ^
/home/andy/misc/zig/src/Sema.zig:597:64: 0x318f5ec in Sema.analyzeBodyInner (zig)
            .as_node                      => try sema.zirAsNode(block, inst),
                                                               ^
```

The ZIR code looks like this:

```
%0 = extended(struct_decl(parent, Auto, {
  [39] export entry line(0) hash(b7894a539990f9a69e05229c0e1265b9): %1 = block_inline({
    %22 = func(ret_ty={
      %2 = break_inline(%22, @Ref.void_type)
    }, body={
      %3 = dbg_stmt(2, 5)
      %4 = alloc_mut(@Ref.bool_type) node_offset:2:5
      %5 = store_node(%4, @Ref.bool_true) node_offset:2:22
      %6 = dbg_stmt(3, 5)
      %11 = block({
        %8 = load(%4) node_offset:3:24
        %9 = as_node(@Ref.bool_type, %8) node_offset:3:24
        %10 = condbr(%9, {
          %12 = int(32768)
          %15 = break(%11, %12)
        }, {
          %16 = break(%11, @Ref.zero)
        }) node_offset:3:20
      }) node_offset:3:20
      %18 = dbg_stmt(4, 5)
      %19 = as_node(@Ref.u16_type, %11) node_offset:4:18
      %20 = ensure_result_non_error(%19) node_offset:4:9
      %21 = ret_coerce(@Ref.void_value) token_offset:5:1
    }) (lbrace=1:24,rbrace=5:1) node_offset:1:8
    %23 = break_inline(%1, %22)
  }) node_offset:1:8
}, {}, {})
```

In particular notice that the `u16` type information is not represented anywhere. This causes a comptime int to be created that is considered to be runtime known. This is causing the crash. Instead, this ZIR should be emitted:

```diff
        %10 = condbr(%9, {
          %12 = int(32768)
+          %coerced = as(@Ref.u16, %12)
          %15 = break(%11, %coerced)
        }, {
+          %coerced = as(@Ref.u16, @Ref.zero)
          %16 = break(%11, %coerced)
        }) node_offset:3:20
```

Here's a clue of relevant code to get started: https://github.com/ziglang/zig/blob/6f87f49f3ddf4fb56f8b4e2ff86d1020ab2dd690/src/AstGen.zig#L2566-L2708

**Comment by schmee** [comment]
I just ran into this while trying to fix another issue, I'll see if I can fix it 👍

**Comment by andrewrk** [comment]
Thanks for the fix @schmee - landed in fd1284ebd07ded1c67bbaff4c14f093051e56f59.
